From ef5954b8e31519e5f496983e753a690e2c10a522 Mon Sep 17 00:00:00 2001
From: Ivan Pavlotsky <ivan.pavlotsky@lge.com>
Date: Fri, 23 Aug 2024 19:27:15 +0300
Subject: [PATCH] 129.0.6654.0~1: Disable Neon DotProduct and I8MM instructions
 in libyuv

:Release Notes:
Disabled Neon DotProduct and I8MM instructions in libyuv to fix
webOS/OSE build after rolling to 129.0.6654.0~1

:Detailed Notes:
In the upstream libyuv CLs:
- https://chromium-review.googlesource.com/c/libyuv/libyuv/+/5383442
- https://chromium-review.googlesource.com/c/libyuv/libyuv/+/5424737
- https://chromium-review.googlesource.com/c/libyuv/libyuv/+/5463945
- https://chromium-review.googlesource.com/c/libyuv/libyuv/+/5490049
- https://chromium-review.googlesource.com/c/libyuv/libyuv/+/5621577
- https://chromium-review.googlesource.com/c/libyuv/libyuv/+/5712928
(which were rolled into Chromium in the upstream CL
https://crrev.com/c/5778746) added and used Neon DotProduct and I8MM
instructions.
It caused errors on building webOS/OSE like the following:
| FAILED: obj/third_party/libyuv/libyuv_neon/compare_neon64.o
...
| {standard input}: Assembler messages:
| {standard input}:154: Error: selected processor does not support `udot v4.4s,v0.16b,v6.16b'
| {standard input}:155: Error: selected processor does not support `udot v5.4s,v1.16b,v6.16b'
| {standard input}:189: Error: selected processor does not support `udot v4.4s,v0.16b,v0.16b'
| {standard input}:190: Error: selected processor does not support `udot v5.4s,v1.16b,v1.16b'

Disable Neon DotProduct and I8MM instructions in libyuv for webOS/OSE to
fix this issue.

:Issues Addressed:
[NEVA-9526] Upgrade neva-main-rolling up to 6654 BUILD
---
 BUILD.gn                     |  7 ++++++-
 include/libyuv/compare_row.h |  4 ++++
 include/libyuv/row.h         |  4 ++++
 source/compare_neon64.cc     |  4 ++++
 source/row_neon64.cc         | 40 ++++++++++++++++++++++++++++++++++++
 5 files changed, 58 insertions(+), 1 deletion(-)

diff --git a/BUILD.gn b/BUILD.gn
index d3fbd33e..42fe1df7 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -7,6 +7,7 @@
 # be found in the AUTHORS file in the root of the source tree.
 
 import("//build/config/features.gni")
+import("//neva/neva.gni")
 import("//testing/test.gni")
 import("libyuv.gni")
 
@@ -230,7 +231,11 @@ if (libyuv_use_neon) {
       use_libcxx_modules = false
 
       # TODO(thakis): Do we get this from global flags already? Is this needed?
-      cflags = [ "-march=armv8-a+dotprod+i8mm" ]
+      # TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+      # instructions.
+      if (!is_webos) {
+        cflags = [ "-march=armv8-a+dotprod+i8mm" ]
+      }
     }
   }
 }
diff --git a/include/libyuv/compare_row.h b/include/libyuv/compare_row.h
index a08734e9..942f6976 100644
--- a/include/libyuv/compare_row.h
+++ b/include/libyuv/compare_row.h
@@ -61,8 +61,12 @@ extern "C" {
 #if !defined(LIBYUV_DISABLE_NEON) && defined(__aarch64__)
 #define HAS_HASHDJB2_NEON
 
+// TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+// instructions.
+#if !defined(OS_WEBOS)
 #define HAS_HAMMINGDISTANCE_NEON_DOTPROD
 #define HAS_SUMSQUAREERROR_NEON_DOTPROD
+#endif  // !defined(OS_WEBOS)
 #endif
 
 uint32_t HammingDistance_C(const uint8_t* src_a,
diff --git a/include/libyuv/row.h b/include/libyuv/row.h
index 569bd6ba..97088120 100644
--- a/include/libyuv/row.h
+++ b/include/libyuv/row.h
@@ -552,6 +552,9 @@ extern "C" {
 #define HAS_P410TOAR30ROW_NEON
 #define HAS_P410TOARGBROW_NEON
 
+// TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+// instructions.
+#if !defined(OS_WEBOS)
 #define HAS_ABGRTOYJROW_NEON_DOTPROD
 #define HAS_ABGRTOYROW_NEON_DOTPROD
 #define HAS_ARGBGRAYROW_NEON_DOTPROD
@@ -565,6 +568,13 @@ extern "C" {
 #define HAS_ARGBCOLORMATRIXROW_NEON_I8MM
 #define HAS_ARGBTOUVMATRIXROW_NEON_I8MM
 #define HAS_ARGBTOUV444MATRIXROW_NEON_I8MM
+#define HAS_ARGBTOUV444ROW_NEON_I8MM
+#define HAS_ARGBTOUVJ444ROW_NEON_I8MM
+#define HAS_ARGBTOUVJROW_NEON_I8MM
+#define HAS_ARGBTOUVROW_NEON_I8MM
+#define HAS_BGRATOUVROW_NEON_I8MM
+#define HAS_RGBATOUVROW_NEON_I8MM
+#endif  // !defined(OS_WEBOS)
 #endif
 
 // The following are available on AArch64 SVE platforms:
diff --git a/source/compare_neon64.cc b/source/compare_neon64.cc
index 36c5e575..6a79cb4b 100644
--- a/source/compare_neon64.cc
+++ b/source/compare_neon64.cc
@@ -153,6 +153,9 @@ uint32_t HashDjb2_NEON(const uint8_t* src, int count, uint32_t seed) {
   return hash;
 }
 
+// TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+// instructions.
+#if !defined(OS_WEBOS)
 uint32_t HammingDistance_NEON_DotProd(const uint8_t* src_a,
                                       const uint8_t* src_b,
                                       int count) {
@@ -214,6 +217,7 @@ uint32_t SumSquareError_NEON_DotProd(const uint8_t* src_a,
       : "memory", "cc", "v0", "v1", "v2", "v3", "v4", "v5");
   return sse;
 }
+#endif  // !defined(OS_WEBOS)
 
 #endif  // !defined(LIBYUV_DISABLE_NEON) && defined(__aarch64__)
 
diff --git a/source/row_neon64.cc b/source/row_neon64.cc
index 782a6653..df51114a 100644
--- a/source/row_neon64.cc
+++ b/source/row_neon64.cc
@@ -2848,11 +2848,14 @@ void RGBToUV444MatrixRow_NEON(const uint8_t* src_rgb,
         "v27");
 }
 
-void ARGBToUV444MatrixRow_NEON_I8MM(const uint8_t* src_argb,
-                                    uint8_t* dst_u,
-                                    uint8_t* dst_v,
-                                    int width,
-                                    const struct ArgbConstants* c) {
+// TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+// instructions.
+#if !defined(OS_WEBOS)
+static void ARGBToUV444MatrixRow_NEON_I8MM(const uint8_t* src_argb,
+                                           uint8_t* dst_u,
+                                           uint8_t* dst_v,
+                                           int width,
+                                           const struct ArgbConstants* c) {
   asm volatile(
       "ldr         q16, [%[c], #16]              \n"  // kRGBToU
       "ldr         q17, [%[c], #32]              \n"  // kRGBToV
@@ -2885,6 +2888,7 @@ void ARGBToUV444MatrixRow_NEON_I8MM(const uint8_t* src_argb,
       : "cc", "memory", "v0", "v1", "v2", "v3", "v4", "v5", "v16", "v17",
         "v29");
 }
+#endif  // !defined(OS_WEBOS)
 
 // RGB to BT601 coefficients
 // UB   0.875 coefficient = 112
@@ -2893,6 +2897,43 @@ void ARGBToUV444MatrixRow_NEON_I8MM(const uint8_t* src_argb,
 // VB -0.1406 coefficient = -18
 // VG -0.7344 coefficient = -94
 // VR   0.875 coefficient = 112
+void ARGBToUV444Row_NEON(const uint8_t* src_argb,
+                         uint8_t* dst_u,
+                         uint8_t* dst_v,
+                         int width) {
+  ARGBToUV444MatrixRow_NEON(src_argb, dst_u, dst_v, width, &kArgbI601Constants);
+}
+
+// TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+// instructions.
+#if !defined(OS_WEBOS)
+void ARGBToUV444Row_NEON_I8MM(const uint8_t* src_argb,
+                              uint8_t* dst_u,
+                              uint8_t* dst_v,
+                              int width) {
+  ARGBToUV444MatrixRow_NEON_I8MM(src_argb, dst_u, dst_v, width,
+                                 &kArgbI601Constants);
+}
+#endif  // !defined(OS_WEBOS)
+
+void ARGBToUVJ444Row_NEON(const uint8_t* src_argb,
+                          uint8_t* dst_u,
+                          uint8_t* dst_v,
+                          int width) {
+  ARGBToUV444MatrixRow_NEON(src_argb, dst_u, dst_v, width, &kArgbJPEGConstants);
+}
+
+// TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+// instructions.
+#if !defined(OS_WEBOS)
+void ARGBToUVJ444Row_NEON_I8MM(const uint8_t* src_argb,
+                               uint8_t* dst_u,
+                               uint8_t* dst_v,
+                               int width) {
+  ARGBToUV444MatrixRow_NEON_I8MM(src_argb, dst_u, dst_v, width,
+                                 &kArgbJPEGConstants);
+}
+#endif  // !defined(OS_WEBOS)
 #define RGBTOUV_SETUP_REG                                                  \
   "movi       v20.8h, #112          \n" /* UB/VR coefficient  (0.875)   */ \
   "movi       v21.8h, #74           \n" /* UG coefficient    (-0.5781)  */ \
@@ -3274,6 +3315,9 @@ void ARGB4444ToUVRow_NEON(const uint8_t* src_argb4444,
   );
 }
 
+// TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+// instructions.
+#if !defined(OS_WEBOS)
 // Process any of ARGB, ABGR, BGRA, RGBA, by adjusting the ArgbConstants layout.
 static void ARGBToUVMatrixRow_NEON_I8MM_Impl(const uint8_t* src,
                                              int src_stride,
@@ -3354,6 +3398,63 @@ void ARGBToUVMatrixRow_NEON_I8MM(const uint8_t* src_argb,
                                    width, c);
 }
 
+// TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+// instructions.
+#if !defined(OS_WEBOS)
+void ARGBToUVRow_NEON_I8MM(const uint8_t* src_argb,
+                           int src_stride_argb,
+                           uint8_t* dst_u,
+                           uint8_t* dst_v,
+                           int width) {
+  ARGBToUVMatrixRow_NEON_I8MM_Impl(src_argb, src_stride_argb, dst_u, dst_v,
+                                   width, &kArgbI601Constants);
+}
+
+void ABGRToUVRow_NEON_I8MM(const uint8_t* src_abgr,
+                           int src_stride_abgr,
+                           uint8_t* dst_u,
+                           uint8_t* dst_v,
+                           int width) {
+  ARGBToUVMatrixRow_NEON_I8MM_Impl(src_abgr, src_stride_abgr, dst_u, dst_v,
+                                   width, &kAbgrI601Constants);
+}
+
+void BGRAToUVRow_NEON_I8MM(const uint8_t* src_bgra,
+                           int src_stride_bgra,
+                           uint8_t* dst_u,
+                           uint8_t* dst_v,
+                           int width) {
+  ARGBToUVMatrixRow_NEON_I8MM_Impl(src_bgra, src_stride_bgra, dst_u, dst_v,
+                                   width, &kBgraI601Constants);
+}
+
+void RGBAToUVRow_NEON_I8MM(const uint8_t* src_rgba,
+                           int src_stride_rgba,
+                           uint8_t* dst_u,
+                           uint8_t* dst_v,
+                           int width) {
+  ARGBToUVMatrixRow_NEON_I8MM_Impl(src_rgba, src_stride_rgba, dst_u, dst_v,
+                                   width, &kRgbaI601Constants);
+}
+
+void ARGBToUVJRow_NEON_I8MM(const uint8_t* src_argb,
+                            int src_stride_argb,
+                            uint8_t* dst_u,
+                            uint8_t* dst_v,
+                            int width) {
+  ARGBToUVMatrixRow_NEON_I8MM_Impl(src_argb, src_stride_argb, dst_u, dst_v,
+                                   width, &kArgbJPEGConstants);
+}
+
+void ABGRToUVJRow_NEON_I8MM(const uint8_t* src_abgr,
+                            int src_stride_abgr,
+                            uint8_t* dst_u,
+                            uint8_t* dst_v,
+                            int width) {
+  ARGBToUVMatrixRow_NEON_I8MM_Impl(src_abgr, src_stride_abgr, dst_u, dst_v,
+                                   width, &kAbgrJPEGConstants);
+}
+#endif  // !defined(OS_WEBOS)
 void RGB565ToYRow_NEON(const uint8_t* src_rgb565, uint8_t* dst_y, int width) {
   asm volatile(
       "movi        v24.16b, #25                  \n"  // B * 0.1016 coefficient
@@ -3487,6 +3588,9 @@ void ARGBToYMatrixRow_NEON(const uint8_t* src_argb,
         "v19", "v20", "v21", "v22");
 }
 
+// TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+// instructions.
+#if !defined(OS_WEBOS)
 void ARGBToYMatrixRow_NEON_DotProd(const uint8_t* src_argb,
                                    uint8_t* dst_y,
                                    int width,
@@ -3520,6 +3624,7 @@ void ARGBToYMatrixRow_NEON_DotProd(const uint8_t* src_argb,
       : "cc", "memory", "v0", "v1", "v2", "v3", "v4", "v5", "v6", "v7", "v16",
         "v17", "v18", "v19");
 }
+#endif  // !defined(OS_WEBOS)
 
 void RGBToYMatrixRow_NEON_DotProd(const uint8_t* src_rgb,
                                   uint8_t* dst_y,
@@ -3565,7 +3670,85 @@ void RGBToYMatrixRow_NEON_DotProd(const uint8_t* src_rgb,
         "v27");
 }
 
+// RGB to JPeg coefficients
+
+void ARGBToYRow_NEON(const uint8_t* src_argb, uint8_t* dst_y, int width) {
+  ARGBToYMatrixRow_NEON(src_argb, dst_y, width, &kArgbI601Constants);
+}
+
+void ARGBToYJRow_NEON(const uint8_t* src_argb, uint8_t* dst_yj, int width) {
+  ARGBToYMatrixRow_NEON(src_argb, dst_yj, width, &kArgbJPEGConstants);
+}
+
+void ABGRToYRow_NEON(const uint8_t* src_abgr, uint8_t* dst_y, int width) {
+  ARGBToYMatrixRow_NEON(src_abgr, dst_y, width, &kAbgrI601Constants);
+}
+
+void ABGRToYJRow_NEON(const uint8_t* src_abgr, uint8_t* dst_yj, int width) {
+  ARGBToYMatrixRow_NEON(src_abgr, dst_yj, width, &kAbgrJPEGConstants);
+}
+
+// TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+// instructions.
+#if !defined(OS_WEBOS)
+void ARGBToYRow_NEON_DotProd(const uint8_t* src_argb,
+                             uint8_t* dst_y,
+                             int width) {
+  ARGBToYMatrixRow_NEON_DotProd(src_argb, dst_y, width, &kArgbI601Constants);
+}
+
+void ARGBToYJRow_NEON_DotProd(const uint8_t* src_argb,
+                              uint8_t* dst_yj,
+                              int width) {
+  ARGBToYMatrixRow_NEON_DotProd(src_argb, dst_yj, width, &kArgbJPEGConstants);
+}
+
+void ABGRToYRow_NEON_DotProd(const uint8_t* src_abgr,
+                             uint8_t* dst_y,
+                             int width) {
+  ARGBToYMatrixRow_NEON_DotProd(src_abgr, dst_y, width, &kAbgrI601Constants);
+}
+
+void ABGRToYJRow_NEON_DotProd(const uint8_t* src_abgr,
+                              uint8_t* dst_yj,
+                              int width) {
+  ARGBToYMatrixRow_NEON_DotProd(src_abgr, dst_yj, width, &kAbgrJPEGConstants);
+#endif  // !defined(OS_WEBOS)
+
 // RGBA expects first value to be A and ignored, then 3 values to contain RGB.
+void RGBAToYRow_NEON(const uint8_t* src_rgba, uint8_t* dst_y, int width) {
+  ARGBToYMatrixRow_NEON(src_rgba, dst_y, width, &kRgbaI601Constants);
+}
+
+void RGBAToYJRow_NEON(const uint8_t* src_rgba, uint8_t* dst_yj, int width) {
+  ARGBToYMatrixRow_NEON(src_rgba, dst_yj, width, &kRgbaJPEGConstants);
+}
+
+void BGRAToYRow_NEON(const uint8_t* src_bgra, uint8_t* dst_y, int width) {
+  ARGBToYMatrixRow_NEON(src_bgra, dst_y, width, &kBgraI601Constants);
+}
+
+// TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+// instructions.
+#if !defined(OS_WEBOS)
+void RGBAToYRow_NEON_DotProd(const uint8_t* src_rgba,
+                             uint8_t* dst_y,
+                             int width) {
+  ARGBToYMatrixRow_NEON_DotProd(src_rgba, dst_y, width, &kRgbaI601Constants);
+}
+
+void RGBAToYJRow_NEON_DotProd(const uint8_t* src_rgba,
+                              uint8_t* dst_yj,
+                              int width) {
+  ARGBToYMatrixRow_NEON_DotProd(src_rgba, dst_yj, width, &kRgbaJPEGConstants);
+}
+
+void BGRAToYRow_NEON_DotProd(const uint8_t* src_bgra,
+                             uint8_t* dst_y,
+                             int width) {
+  ARGBToYMatrixRow_NEON_DotProd(src_bgra, dst_y, width, &kBgraI601Constants);
+}
+#endif  // !defined(OS_WEBOS)
 void RGBToYMatrixRow_NEON(const uint8_t* src_rgb,
                           uint8_t* dst_y,
                           int width,
@@ -4020,6 +4203,9 @@ static const uvec8 kARGBGrayRowCoeffs = {29, 150, 77, 0};
 static const uvec8 kARGBGrayRowIndices = {0, 0, 0, 19, 2, 2, 2, 23,
                                           4, 4, 4, 27, 6, 6, 6, 31};
 
+// TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+// instructions.
+#if !defined(OS_WEBOS)
 void ARGBGrayRow_NEON_DotProd(const uint8_t* src_argb,
                               uint8_t* dst_argb,
                               int width) {
@@ -4047,6 +4233,7 @@ void ARGBGrayRow_NEON_DotProd(const uint8_t* src_argb,
         [indices] "r"(&kARGBGrayRowIndices)  // %[indices]
       : "cc", "memory", "v0", "v1", "v2", "v3", "v24", "v25");
 }
+#endif  // !defined(OS_WEBOS)
 
 // Convert 8 ARGB pixels (32 bytes) to 8 Sepia ARGB pixels.
 //    b = (r * 35 + g * 68 + b * 17) >> 7
@@ -4093,6 +4280,9 @@ static const uvec8 kARGBSepiaRowCoeffs = {17, 68, 35, 0,  22, 88,
                                           45, 0,  24, 98, 50, 0};
 static const uvec8 kARGBSepiaRowAlphaIndices = {3, 7, 11, 15, 19, 23, 27, 31};
 
+// TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+// instructions.
+#if !defined(OS_WEBOS)
 void ARGBSepiaRow_NEON_DotProd(uint8_t* dst_argb, int width) {
   asm volatile(
       "ld3r        {v20.4s, v21.4s, v22.4s}, [%[coeffs]] \n"
@@ -4129,6 +4319,7 @@ void ARGBSepiaRow_NEON_DotProd(uint8_t* dst_argb, int width) {
       : "cc", "memory", "v0", "v1", "v2", "v3", "v4", "v5", "v6", "v7", "v20",
         "v21", "v22", "v24", "v25", "v26", "v28", "v29", "v30");
 }
+#endif  // !defined(OS_WEBOS)
 
 // Tranform 8 ARGB pixels (32 bytes) with color matrix.
 // TODO(fbarchard): Was same as Sepia except matrix is provided.  This function
@@ -4192,6 +4383,9 @@ void ARGBColorMatrixRow_NEON(const uint8_t* src_argb,
         "v17", "v18", "v19", "v22", "v23", "v24", "v25");
 }
 
+// TODO(neva): Remove this when webOS supports Neon DotProduct and I8MM
+// instructions.
+#if !defined(OS_WEBOS)
 void ARGBColorMatrixRow_NEON_I8MM(const uint8_t* src_argb,
                                   uint8_t* dst_argb,
                                   const int8_t* matrix_argb,
@@ -4247,6 +4441,7 @@ void ARGBColorMatrixRow_NEON_I8MM(const uint8_t* src_argb,
       : "cc", "memory", "v0", "v1", "v16", "v17", "v18", "v19", "v20", "v21",
         "v22", "v23", "v31");
 }
+#endif  // !defined(OS_WEBOS)
 
 // Multiply 2 rows of ARGB pixels together, 8 pixels at a time.
 void ARGBMultiplyRow_NEON(const uint8_t* src_argb,
